### PR TITLE
Refactor matchers as classes

### PR DESCRIPTION
**Pull request by tagliala**
@chrisalley if you want to take an overall review

I would check messages and, if you are using it actively, test this branch against a large codebase

### CODE REVIEW

**Review by chrisalley — CHANGES_REQUESTED** [review]
I've tested this branch against a codebase with 392 policy tests. There were no failures. (Note: the specs don't use `permit_mass_assignment_of` extensively, and do not use `not_to` since that was never supported.)

Overall I like your approach. The internals have been simplified  and the use of "helper matchers" reduces the number of files that we need to have.

Great job!

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,32 @@
...
+The following matchers may fail, because there were ambiguous in the previous
+implementation, and the application policies should be fixed:
+- `not_to forbid_edit_and_update_actions`
+- `not_to forbid_new_and_create_actions`
+- `not_to forbid_actions` (with multiple actions)
+- `not_to forbid_mass_assignment_of` (with multiple attributes)
+- `not_to permit_actions` (with multiple actions)
+- `not_to permit_edit_and_update_actions`
+- `not_to permit_new_and_create_actions`
+- `not_to permit_mass_assignment_of` (with multiple attributes)
```
Comment (tagliala) [inline CHANGELOG.md:31]: new, to review

I've also eliminated a breaking change for #35 (I will do it separately with better support for specific aliases)

File: lib/pundit/matchers/base_matcher.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+require_relative 'utils/policy_info'
+
+module Pundit
+  module Matchers
+    # This is the base class for all matchers in the Pundit Matchers library.
+    class BaseMatcher
+      # Error message when an ambiguous matcher is used.
+      AMBIGUOUS_MATCHER_ERROR = <<~MSG
+        `expect().%<matcher>s` is not supported since it creates ambiguity.
+        Instead use `expect().%<suggested_matcher>s`'
```
Comment (tagliala) [inline lib/pundit/matchers/base_matcher.rb:12]: I'll remove the suggestion because no one can understand what `not_to forbid_all_actions` means

**Comment by chrisalley** [comment]
There's some solid work here; much appreciated. Since this is already a big set of changes, it would be good to have the branch finalised for a 3.0.0 release. It can be used as the basis for smaller releases in the future.

**Comment by tagliala** [comment]
If you can merge this, I will incrementally

- Add support for multiple user aliases (easy **breaking change** in case of `user_alias` customization) #35 
- Add support for nested attributes #7

**Pull request merged**